notes/books/philosophy-software-design.md

```diff
@@ -335,9 +335,245 @@ subjective.
+Every company has one, and one of the most important cultural changes a
+company can make is to not reward that kind of behavior.
+
+The strategic programmer on the other hand is the opposite, he wants
+to deliver software as fast as possible, but not in detriment of the
+future, so he takes time to address design issues and fix them, even if
+it means more work now. He also spends more time experimenting with
+different designs to solve a problem, so he can choose the design
+that fits best the problem, because just making it work is not enough.
+
+
+## Deep modules
+
+Before going in this topic it is important to define what is
+a module in this context. Here a module can be an object,
+a package, or a service. It is a unit of code that is cohesive
+and has a public interface, be it local or remote.
+
+This is one of the most interesting chapters of the book since it
+gives advice that is very sound and yet goes against the current
+fad of micro-services. Instead of using being small as a way to
+define that a module is good, it uses the idea of being deep, which
+is the opposite.
+
+In the opinion of the author, good modules are deep, in the sense that
+they have small and simple interfaces (the interface is not very wide),
+which hides a lot of complexity underneath (hence deep).
+
+Usually breaking up a module just because
+its implementation is big will only spread the complexity around,
+which the maintainer will need to understand anyway, and introduce
+more interfaces, so instead of having just one simple interface you
+may end up having N interfaces to smaller implementation. Which is
+the situation that gives birth to ideas like aggregation layers, 
+because composing all this small interfaces is complex,
+so you build the interface you wanted all the way along on top of it.
+
+In the end you cause more trouble to the clients
+of your module just to satisfy your desire for small modules. This
+will also cause trouble to maintainers too, since you will end up
+jumping back and forth through a lot of small/shallow modules to
+understand even basic things about the system.
+
+His examples of shallow modules are all related to Java, which is a nice
+example of the mindset of "objects are good, so lets have a lot of them".
+Even to do simple tasks you end up having to create multiple objects
+and composing them, which adds complexity to clients. This made me
+remember how I felt when I wrote a program to read a file contents in Python
+and compared it with Java, it made me shift to Python for all my
+work in college and I never looked back. So you have all this micro/small
+everything thing that kinda makes sense, but the end result is just
+terrible to work with (always composing one trillion small things).
+
+In the case of micro-service this is even worse, because composing
+and understanding multiple services is harder than local objects.
+So the whole movement gives me bad feelings,
```

Comment:
You're not wrong, but I'd call myself an advocate of microservices so I feel like I need to offer an alternative view :smile_cat:

In my mind, "microservices" means two very different things to different people:

1. A style of software design that composes a system of a set of collaborating services, and that tries to avoid the pitfalls of some of the earlier approaches such as "web services". There are some guidelines for services -- one of them is "services should be independently deployable" -- but it doesn't say much about the size of services.
2. The rule that you should build your system out of services that are micro in a literal sense, following guidelines like "no service should be more than 1k lines of code" or "you should be able to rewrite any service in a week."

If we go with the second definition, I completely agree with your point.

Reply:
Yeah I agree with you, but then the name is not very good because it has "micro" on it =P. What you described is basically a service oriented architecture done right to me, I never got the micro part, I know in the ended it was added just to try to avoid the errors of the past (for some people, others go with the size thing), but still it feels off and confusing because answering what is the difference between SOA and microservices does not revolve around the actual size of the services (but the name implies that).